crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Value.hs

```diff
@@ -133,3 +140,125 @@ instance IsExpr (SymExpr sym) => Show (LLVMVal sym) where
     unwords $ [ "[" ]
            ++ intersperse ", " (map show $ V.toList xs)
            ++ [ "]" ]
+
+-- | An efficient n-way @and@: it quits early if it finds any concretely false
+-- values, rather than chaining a bunch of 'andPred's.
+allOf :: (IsExprBuilder sym)
+      => sym
+      -> [Pred sym]
+      -> IO (Pred sym)
+allOf sym xs =
+  if and (mapMaybe asConstantPred xs)
+  then foldM (andPred sym) (truePred sym) xs
+  else pure (falsePred sym)
+
```

Comment:
Are you confident that this is more efficient than `andAllOf` from What4.Interface?  `andPred` itself already has constant-folding behavior...

Reply:
Consider the list `[<symbolic>, <symbolic>, ..., <concretely-true>]`. `andAllOf` will have to apply `andPred` _n_-1 times to symbolic predicates before returning true, whereas this will return true immediately.

I guess it depends on how cheap of an operation `andPred` is. I don't really know much about its implementation. 

It might be best to use `andAllOf` in the `then` branch here, to reduce duplication.